File: DAY06/P게임맵최단거리/P게임맵최단거리_서은효.java
package P게임맵최단거리;

import java.io.IOException;
import java.util.ArrayDeque;
import java.util.LinkedList;
import java.util.Queue;

public class P게임맵최단거리_서은효 {

	static boolean isVistied[][];
	static int [] dx = {1,0,-1,0};
	static int [] dy = {0,1,0-1};
	
	
	public static boolean check(int x, int y) {
		if(x<0 || x>=5 || y<0 || y>=5) return false;
		else return true;
	}
	
	public static void BFS(int x, int y, int [][] map) {
		
		Queue<int[]> que = new LinkedList<>();
		que.add(new int [] {x,y});
	
		while(!que.isEmpty()){
			int[] now = que.poll();
			int nowx = now[0];
			int nowy = now[1];
			
			for(int i=0; i<4; i++) {
				int next_x = nowx + dx[i];
				int next_y = nowy + dy[i];
				
				if(!check(next_x, next_y)
						|| map[next_x][next_y] == 0
						|| isVistied[next_x][next_y]) {
					continue;
				}
				que.add(new int[] {next_x, next_y});
				map[next_x][next_y] = map[now[0]][now[1]] +1;
				isVistied[next_x][next_y] = true;
			}
			

			
		}
		
		
		
	}
	
    public static int solution(int[][] maps) {
        int answer = 0;
        
    	isVistied = new boolean [5][5];
    	isVistied[0][0] = true;
    	
    	BFS(0,0, maps);
    	answer = maps[4][4];
        return answer;
        
        
    }
    public static void main(String[] args) throws IOException {
    	int [][] map = {{1,0,1,1,1},{1,0,1,0,1},{1,0,1,1,1},{1,1,1,0,1},{0,0,0,0,1}};
    	int result = solution(map);
    	System.out.println(result);
    }
}

File: DAY06/P게임맵최단거리/P게임맵최단거리_서은효.py
from collections import deque


def BFS(n,m,maps):
    dx = [-1,1,0,0]
    dy = [0,0,-1,1]
    queue = deque()
    queue.append((0,0))#처음 시작

    while queue:
        x,y = queue.popleft()
        for d in range(4):
            nx = x + dx[d]
            ny = y + dy[d]

            if 0<=nx<n and 0<=ny<m and maps[nx][ny]==1: #검증
                queue.append((nx,ny)) #큐에 다시 저장
                maps[nx][ny] = maps[x][y] + 1 #DP

    if maps[n-1][m-1] == 1: #만약 n-1,m-1(목표지)가 1인 경우
        return -1 #목표지가 1이라면 목표지까지 더해가며 도달하지 못했다는 뜻
    else:
        return maps[n-1][m-1] #그렇지 않다면 더해온 값을 리턴 

def solution(maps):
    n = len(maps) #행
    m = len(maps[0]) #열
    answer = BFS(n,m,maps) #배열의 크기와 배열
    return answer


print(solution([[1,0,1,1,1],[1,0,1,0,1],[1,0,1,1,1],[1,1,1,0,1],[0,0,0,0,1]]))
